edx_notifications/base_data.py

```diff
@@ -0,0 +1,287 @@
+"""
+Base objects that data.py uses
+"""
+
+import json
+import copy
+import inspect
+import dateutil.parser
+
+from datetime import datetime
+
+from weakref import WeakKeyDictionary
+
+
+class TypedField(object):
+    """
+    Field Decscriptors used to enforce correct typing
+    """
+
+    _expected_type = None
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initializer which takes in the type this field
+        should be set it is set
+        """
+
+        self._expected_type = [args[0]] if not isinstance(args[0], list) else args[0]
+        self._default = kwargs.get('default', None)
+        self._data = WeakKeyDictionary()
```

Comment:
In `__set__`, you have the instance object. You could just do something like: `instance.__data[self] = value`.